### PR TITLE
Rewrite `USING` to `ON` condition for joins

### PR DESCRIPTION
**Pull request by frouioui**
## Description

This PR rewrites `USING` condition to `ON` instead of `WHERE` to avoid the issue mentioned in https://github.com/vitessio/vitess/issues/13929.

## Related Issue(s)

- Fixes https://github.com/vitessio/vitess/issues/13929

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on the CI
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vtgate/semantics/early_rewriter.go [hunk 1]
```diff
@@ -344,34 +344,13 @@ func rewriteOrFalse(orExpr sqlparser.OrExpr) sqlparser.Expr {
...
-			sel.Where = nil
-		}
-		sel.Where = &sqlparser.Where{
-			Type: sqlparser.WhereClause,
-			Expr: sqlparser.AndExpressions(predicates...),
-		}
-		return nil
-	}
-	return vterrors.Errorf(vtrpcpb.Code_INTERNAL, "did not find WHERE clause")
+	cond.On = sqlparser.AndExpressions(predicates...)
```
Comment (harshit-gangal) [inline go/vt/vtgate/semantics/early_rewriter.go:352]: This is only problematic with `outer join` queries that an `on` condition is required. With `inner join` this is not the case. Should we not check and do rewriting based on that.

File: go/vt/vtgate/semantics/early_rewriter_test.go [hunk 2]
```diff
@@ -146,24 +146,24 @@ func TestExpandStar(t *testing.T) {
...
+		expSQL:   "select t2.c1 as c1, t2.c2 as c2, t4.c4 as c4 from t2 join t4 on t2.c1 = t4.c1",
 		expanded: "main.t2.c1, main.t2.c2, main.t4.c4",
 	}, {
 		sql:    "select * from t2 join t4 using (c1) join t2 as X using (c1)",
-		expSQL: "select t2.c1 as c1, t2.c2 as c2, t4.c4 as c4, X.c2 as c2 from t2 join t4 join t2 as X where t2.c1 = t4.c1 and t2.c1 = X.c1 and t4.c1 = X.c1",
+		expSQL: "select t2.c1 as c1, t2.c2 as c2, t4.c4 as c4, X.c2 as c2 from t2 join t4 on t2.c1 = t4.c1 join t2 as X on t2.c1 = t4.c1 and t2.c1 = X.c1 and t4.c1 = X.c1",
 	}, {
 		sql:    "select * from t2 join t4 using (c1), t2 as t2b join t4 as t4b using (c1)",
-		expSQL: "select t2.c1 as c1, t2.c2 as c2, t4.c4 as c4, t2b.c1 as c1, t2b.c2 as c2, t4b.c4 as c4 from t2 join t4, t2 as t2b join t4 as t4b where t2b.c1 = t4b.c1 and t2.c1 = t4.c1",
+		expSQL: "select t2.c1 as c1, t2.c2 as c2, t4.c4 as c4, t2b.c1 as c1, t2b.c2 as c2, t4b.c4 as c4 from t2 join t4 on t2.c1 = t4.c1, t2 as t2b join t4 as t4b on t2b.c1 = t4b.c1",
```
Comment (harshit-gangal) [inline go/vt/vtgate/semantics/early_rewriter_test.go:156]: add a `left join` and `right join` case here.

File: go/vt/vtgate/semantics/binder.go [hunk 3]
```diff
@@ -85,7 +85,23 @@ func (b *binder) up(cursor *sqlparser.Cursor) error {
...
+					return true, nil
+				}
+				deps, err := b.resolveColumn(column, currScope, false)
+				if err != nil {
+					return false, err
+				}
+				b.recursive[column] = deps.recursive
+				b.direct[column] = deps.direct
+				return true, nil
+			}, node.On)
```
Comment (harshit-gangal) [inline go/vt/vtgate/semantics/binder.go:104]:
```
join_specification: {
    ON search_condition
  | USING (join_column_list)
}

join_column_list:
    column_name [, column_name] ...
```

`Using` can only have `column_name`, I think we can skip tree walking and can speed up the dependency assignment.

Comment (harshit-gangal) [reply go/vt/vtgate/semantics/binder.go:104]: Also `resolveColumn` call is not needed, when `Using` join spec is used, the column should belong to both left and right tableExpr. Probably can get that information directly and assign it.

File: go/vt/vtgate/semantics/early_rewriter.go (hunk 1 above)
Comment (systay) [reply go/vt/vtgate/semantics/early_rewriter.go:352]: I think it's cleaner and simpler to just always move it to the `on` clause. the rest of the planner doesn't care where the predicates are for inner joins anyway

Comment (frouioui) [reply go/vt/vtgate/semantics/early_rewriter.go:352]: i agree with @systay

File: go/vt/vtgate/semantics/early_rewriter_test.go (hunk 2 above)
Comment (frouioui) [reply go/vt/vtgate/semantics/early_rewriter_test.go:162]: resolved via 4faa0def6d

File: go/vt/vtgate/semantics/binder.go (hunk 3 above)
Comment (frouioui) [reply go/vt/vtgate/semantics/binder.go:104]: > `Using` can only have `column_name`, I think we can skip tree walking and can speed up the dependency assignment.

I am not sure I understand this correctly. I am walking on `node.On` which is a sqlparser expression. My thinking was that I wanted to catch all columns in this expression and assign dependency on it.

**Review by harshit-gangal — APPROVED** [review]

File: go/vt/schemadiff/schema_test.go [hunk 4]
```diff
@@ -631,7 +631,7 @@ func TestViewReferences(t *testing.T) {
 				"create table t2(id int primary key, n int, info int)",
 				"create view v1 as select id, c as ch from t1 where id > 0",
 				"create view v2 as select n as num, info from t2",
-				"create view v3 as select num, v1.id, ch from v1 join v2 using (id) where info > 5",
+				"create view v3 as select num, v1.id, ch from v1 join v2 on v1.id = v2.num where info > 5",
```
Comment (systay) [inline go/vt/schemadiff/schema_test.go:634]: why is this needed?

File: go/vt/vtgate/planbuilder/testdata/unsupported_cases.json [hunk 5]
```diff
@@ -37,12 +37,12 @@
   {
     "comment": "join with USING construct",
     "query": "select * from user join user_extra using(id)",
-    "plan": "can't handle JOIN USING without authoritative tables"
+    "plan": "VT03019: column id not found"
```
Comment (systay) [inline go/vt/vtgate/planbuilder/testdata/unsupported_cases.json:40]: this doesn't seem correct. we don't know that the column doesn't exist, we just don't have enough info to assert it, right? shouldn't we keep the old error message here?

**Review by systay — APPROVED** [review]

File: go/vt/vtgate/planbuilder/testdata/unsupported_cases.json (hunk 5 above)
Comment (frouioui) [reply go/vt/vtgate/planbuilder/testdata/unsupported_cases.json]: solved via a07c5dfbc0

File: go/vt/schemadiff/schema_test.go (hunk 4 above)
Comment (frouioui) [reply go/vt/schemadiff/schema_test.go:634]: This needed as the original `using` clause was illegal. We are now catching this and erroring out. View `v1` has column `id` but `v2` does not have it, leading to error. I rewrote this test to using `on` to have a working query.
